NIFI-7253 Standardized Jackson dependencies across the codebase.

MikeThomsen opened the pull request:
<!--
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

exceptionfactory commented:
Thanks for putting together this improvement @MikeThomsen! Jackson 2 compatibility has remained solid for a number of releases, so moving to a common version seems like a good solution. As an adjustment to the approach, it seems like it would be better to use the `jackson-bom` dependency to ensure that all related Jackson 2 libraries have the same version. That also streamlines the dependency declarations in the root Maven configuration.

As part of this change, it also looks like the `jackson.version` property could be removed from a number of modules.

exceptionfactory commented on pom.xml:
```diff
@@ -551,6 +551,28 @@
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+<!--            <dependency>-->
+<!--                <groupId>com.fasterxml.jackson.core</groupId>-->
+<!--                <artifactId>jackson-databind</artifactId>-->
+<!--                <version>${jackson.version}</version>-->
+<!--            </dependency>-->
+<!--            <dependency>-->
+<!--                <groupId>com.fasterxml.jackson.core</groupId>-->
+<!--                <artifactId>jackson-core</artifactId>-->
+<!--                <version>${jackson.version}</version>-->
+<!--            </dependency>-->
+<!--            <dependency>-->
+<!--                <groupId>com.fasterxml.jackson.core</groupId>-->
+<!--                <artifactId>jackson-annotations</artifactId>-->
+<!--                <version>${jackson.version}</version>-->
+<!--            </dependency>-->
```
These commented lines should be removed in the final version.

exceptionfactory commented on nifi-commons/nifi-property-protection-factory/pom.xml:
```diff
@@ -21,6 +21,11 @@
         <version>1.16.0-SNAPSHOT</version>
     </parent>
     <artifactId>nifi-property-protection-factory</artifactId>
+    <properties>
+        <!-- We appear to not be able to use the same version of Jackson here because of a subtle bug related to it
+        in the Azure Identity lib-->
+        <jackson.version>2.12.4</jackson.version>
+    </properties>
```
I noticed the following error on the previous build:
```
javax.xml.stream.FactoryConfigurationError: Provider for javax.xml.stream.XMLInputFactory cannot be found
```

Can the Azure library be upgraded, instead of hard-coding a particular Jackson version?

MikeThomsen replied:
I added `woodstox` to satisfy an equivalent issue in the Registry. I'll try that here.

MikeThomsen commented:
@exceptionfactory I addressed both of your concerns. Added `woodstox` in as a replacement for the STaX impl and it removed the build error in the Azure lib.

MikeThomsen commented:
And it would appear that the game of whack-a-mole continues...

MikeThomsen commented:
I think the problem may have started with importing the BOM.

MikeThomsen commented:
Closing for now.
